gtk: add support for using GTK Builder UI files and Blueprints

jcollie opened the pull request:
Adds buildtime and comptime checks to make sure that Blueprints/UI files are availble and correctly formed. Will also compile Blueprints to UI files so that they are available to GTK code.

tristan957 commented on src/apprt/gtk/builder_check.zig:
```diff
@@ -0,0 +1,43 @@
+    defer alloc.free(data);
+
+    if ((comptime !build_options.adwaita) and std.mem.indexOf(u8, data, "lib=\"Adw\"") != null) {
+        std.debug.print("{s}: skipping builder check because Adwaita is not enabled!\n", .{filename});
```
Maybe it would be useful to print the filename/path here?

jcollie replied:
I am, aren't I?

tristan957 commented on src/apprt/gtk/builder_check.zig:
```diff
@@ -0,0 +1,43 @@
+pub const c = @cImport({
+    @cInclude("gtk/gtk.h");
+    if (build_options.adwaita) {
+        @cInclude("libadwaita-1/adwaita.h");
```
I don't understand why Zig is forcing us to include the `libadwaita-1` part of the path here. In C, its `adwaita.h`

jcollie replied:
Must be something left over from previous eras because it no longer seems to be necessary, I was just cargo-culting what was in `gtk/c.zig`.

tristan957 commented on src/build/SharedDeps.zig:
```diff
@@ -493,6 +493,26 @@ pub fn add(
 
+                    {
+                        const builder_check = b.addExecutable(.{
+                            .name = "builder_check",
```
maybe `gtk_builder_check`

jcollie replied:
Done.

pluiedev reviewed (commented):
a) I wonder if zig-gobject would make this a bit better
b) Should we integrate [Blueprint](https://jwestman.pages.gitlab.gnome.org/blueprint-compiler/) as well? Does anyone actually like writing XML?

jcollie commented:
> a) I wonder if zig-gobject would make this a bit better 

`zig-gobject` would likely make a lot of things better, but I wanted to get this out there first.

> b) Should we integrate [Blueprint](https://jwestman.pages.gitlab.gnome.org/blueprint-compiler/) as well? Does anyone actually like writing XML?

I've been thinking about the best way to do so. It'll take some restructuring to get the Blueprint compiler into the build path so that we don't have to keep the generated `.ui` files around.

tristan957 commented:
Yes, blueprint would be great. The language server alone is more than enough reason to use it.

mitchellh commented:
> b) Should we integrate [Blueprint](https://jwestman.pages.gitlab.gnome.org/blueprint-compiler/) as well? Does anyone actually like writing XML?

Supportive.

Just note what I noted in your other PR: if we can integrate compilation into the build pipeline that'd be swell.

jcollie commented:
> > b) Should we integrate [Blueprint](https://jwestman.pages.gitlab.gnome.org/blueprint-compiler/) as well? Does anyone actually like writing XML?
> 
> Supportive.
> 
> Just note what I noted in your other PR: if we can integrate compilation into the build pipeline that'd be swell.

Yeah, I've got generating `.ui` files from `.blp` files 99% figured out. Just polishing the details.

jcollie committed:
gtk: add support for using GTK Builder UI files and Blueprints

Adds buildtime and comptime checks to make sure that Blueprints/UI files
are availble and correctly formed. Will also compile Blueprints to UI
files so that they are available to GTK code.

jcollie commented:
OK, added support for compiling blueprints _and_ using `zig-gobject`.

jcollie commented on src/apprt/gtk/builder_check.zig:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    const builder = c.gtk_builder_new_from_string(data.ptr, @intCast(data.len));
+    defer c.g_object_unref(builder);
```
Yes we can.

mitchellh commented:
I haven't had a chance to look closely (and I don't need to, I defer to you all), but please make sure to review the `build.zig` step dependencies carefully to ensure we didn't introduce any races. 😄 It's a very common mistake and one I've made plenty of times.

pluiedev reviewed (approved):
My zig-build-fu is pretty bad so I can't comment on that but the rest looks pretty good
